### DISCUSSION TITLE
Ring buffer cleanup

### DISCUSSION THREAD

**Discussion by bateyejoe** [axboe/liburing#914] [Q&A]
If I register a ring buffer, fill it with buffers, and use it until I cancel my multi-shot receive, what's the typical way to reclaim the un-used buffers from the buffer ring? I see that I can un-register the buffer ring from the io_uring, but I don't see anything for accessing the unused buffers left in the ring.

**Comment by axboe** [comment]
A dirty approach would be to simply modify the buffer ring index from the application, rewinding it. If you have no pending requests that'd use it, this would work.

A better approach would be to add an API to release N buffers from a given mapped ring, as that could be done consistently and with pending IO as well. We don't have that, but adding it should be pretty easy. Would imagine the input would be which group ID, and how many buffers to "consume" from there. We'd probably want to post a CQE for each one, so the application would know which ones have been freed. Suggestions welcome!
